### ISSUE TITLE
undefined-variable on correct return type hints

### ISSUE THREAD

**Issue by craynic** [PyCQA/pylint#3461]
### Steps to reproduce
code:
```python
from enum import Enum


class Right:
    class Result1(Enum):
        OK = 0
    def work(self) -> Result1:
        return self.Result1.OK


class Wrong:
    class Result2(Enum):
        OK = 0
    def work(self) -> self.Result2:
        return self.Result2.OK
```
Got an error in running:
```
Traceback (most recent call last):
  File "tmp.py", line 11, in <module>
    class Wrong:
  File "tmp.py", line 14, in Wrong
    def work(self) -> self.Result2:
NameError: name 'self' is not defined
```
### Current behavior

tmp.py:7:22: E0602: Undefined variable 'Result1' (undefined-variable)

### Expected behavior

tmp.py:14:22: E0602: Undefined variable 'self' (undefined-variable)

### pylint --version output
pylint 2.4.4
astroid 2.3.3
Python 3.8.2 (default, Feb 26 2020, 22:21:03)

**Comment by v7a** [comment]
The "right" code isn't really correct. You'd need to forward the reference: `def work(self) -> "Result1":`.

**Comment by craynic** [comment]
> The "right" code isn't really correct. You'd need to forward the reference: `def work(self) -> "Result1":`.

I could run it with `python` with no problem. So that's correct.

If that's really not good enough, the message level should be `W`, not `E`.

**Comment by craynic** [comment]
any update?

**Comment by v7a** [comment]
> I could run it with python with no problem. So that's correct.

You're right, my bad, mixed something up. It appears that pylint cannot handle type hints very well. `def work(self) -> "Result1":` seems to satisfy pylint but so does `def work(self) -> "FooBar":` which it shouldn't, of course.

**Comment by PCManticore** [comment]
Thanks for the report. This is definitely a bug, and it's unrelated to type hints. It's caused by `pylint` not properly scoping annotations in the enclosing scope. 

For instance, if the code would be written like this, the error would not be emitted:
```
from enum import Enum

class Result1(Enum):
   OK = 0

class Right:
    def work(self) -> Result1:
        return Result1.OK
```

**Comment by craynic** [comment]
It seems to have been fixed in 2.5. Is it right?
If yes, hope a link to the PR or commit could be given.

**Comment by svenpanne** [comment]
No, this has not been fixed yet. I've just tried to convert some parts of our code base with  https://github.com/ilevkivskyi/com2ann and stumbled over quite a few places which had to be fixed by hand (adding quotes) because pylint doesn't get the scoping right.

It is quite common that signatures of methods in a class Foo involve, well, Foo, so this pylint bug is really annoying in a large code base...

**Comment by svenpanne** [comment]
One more point: pylint should somehow treat typing.TYPE_CHECKING like mypy does, otherwise the scoping is screwed up for another reason.

**Comment by anjsimmo** [comment]
> It seems to have been fixed in 2.5. Is it right?
If yes, hope a link to the PR or commit could be given.

@craynic Yes, the false positive for `Result1` that you reported was fixed by PR #3497 and released in 2.5.0. I've submitted PR #3713 which should also fix the false negative for `self.Result2` once merged.

(If converting a large code base, Pylint may encounter other type hint and scoping related issues--some of which have bug reports--but I think the specific case in this issue is now addressed)
